### PR TITLE
Implement PNI

### PR DESCRIPTION
**Pull request by gferon**
Make use of https://github.com/whisperfish/libsignal-service-rs/pull/285

### CODE REVIEW

File: presage/src/manager/confirmation.rs [hunk 1]
```diff
@@ -87,7 +88,50 @@ impl<S: Store> Manager<S, Confirmation> {
...
+        let (_aci_pre_keys, aci_signed_pre_key, _aci_pq_pre_keys, aci_pq_last_resort_pre_key) =
+            libsignal_service::pre_keys::replenish_pre_keys(
+                &mut self.store,
+                &aci_identity_key_pair,
+                &mut rng,
+                true,
+                0,
+                0,
+            )
+            .await?;
```
Comment (rubdos) [inline presage/src/manager/confirmation.rs:103]: You are aware this block is in AccountManager::register_account right?

Comment (rubdos) [reply presage/src/manager/confirmation.rs:103]: I meant all the way up to the activation request, but scrolling took ages.

Comment (gferon) [reply presage/src/manager/confirmation.rs:103]: Yes, this is the part that's currently broken anyways. I was gonna work on making sure we have a functional place for that, I don't know why it's duplicated anymore.

File: presage/src/manager/registered.rs [hunk 2]
```diff
@@ -953,7 +967,12 @@ impl<S: Store> Manager<S, Registered> {
                         key: profile_key.derive_access_key().to_vec(),
                         certificate: sender_certificate.clone(),
                     });
-            recipients.push((member.uuid.into(), unidentified_access));
+            let include_pni_signature = true;
```
Comment (rubdos) [inline presage/src/manager/registered.rs:970]: This should come from the database, and PNI sigs should only be generated when necessary.

File: presage/src/manager/confirmation.rs (hunk 1 above)
Comment (gferon) [reply presage/src/manager/confirmation.rs]: Removed from `presage` 🥳

**Review by flamingradian — COMMENTED** [review]
Hi, I'm just leaving comments here after testing registration. I'm not a maintainer but would like to see presage registration working again.

Server setup:

    WORK_DIR $ git clone https://github.com/signalapp/Signal-Server
    WORK_DIR $ cd Signal-Server
    WORK_DIR/Signal-Server $ git checkout v12.51.0
    WORK_DIR/Signal-Server $ ./mvnw integration-test -Ptest-server -Dskip-tests=true

Client test:

    WORK_DIR $ git clone https://github.com/whisperfish/libsignal-service-rs
    WORK_DIR $ cd libsignal-service-rs
    WORK_DIR/libsignal-service-rs $ git apply PATCH
    WORK_DIR/libsignal-service-rs $ cd ..
    WORK_DIR $ git clone https://github.com/whisperfish/presage -b aci-pni
    WORK_DIR $ cd presage/presage-cli
    WORK_DIR/presage/presage-cli $ cargo run -- register -s staging --phone-number +xxxxxxxxxxx --captcha signalcaptcha://test.test.registration.test

Patch to libsignal-service-rs that uses the local server:

    diff --git a/libsignal-service/src/configuration.rs b/libsignal-service/src/configuration.rs
    index fff3cc6eb..1cdc5794c 100644
    --- a/libsignal-service/src/configuration.rs
    +++ b/libsignal-service/src/configuration.rs
    @@ -119,16 +119,16 @@ impl From<&SignalServers> for ServiceConfiguration {
                 // configuration with the Signal API staging endpoints
                 // see: https://github.com/signalapp/Signal-Desktop/blob/master/config/default.json
                 SignalServers::Staging => ServiceConfiguration {
    -                service_url: "https://chat.staging.signal.org".parse().unwrap(),
    -                storage_url:"https://storage-staging.signal.org".parse().unwrap(),
    +                service_url: "http://localhost:8080".parse().unwrap(),
    +                storage_url:"https://0.0.0.0".parse().unwrap(),
                     cdn_urls: {
                         let mut map = HashMap::new();
    -                    map.insert(0, "https://cdn-staging.signal.org".parse().unwrap());
    -                    map.insert(2, "https://cdn2-staging.signal.org".parse().unwrap());
    +                    map.insert(0, "https://0.0.0.0".parse().unwrap());
    +                    map.insert(2, "https://0.0.0.0".parse().unwrap());
                         map
                     },
                     contact_discovery_url:
    -                    "https://api-staging.directory.signal.org".parse().unwrap(),
    +                    "https://0.0.0.0".parse().unwrap(),
                     certificate_authority: include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/certs/staging-root-ca.pem")).to_string(),
                     unidentified_sender_trust_root:
                         PublicKey::deserialize(&BASE64_RELAXED.decode("BbqY1DzohE4NUZoVF+L18oUPrK3kILllLEJh2UnPSsEx").unwrap()).unwrap(),

**Comment by boxdot** [comment]
@flamingradian Thank you for testing. Since @gferon is not available right now, I will try to finish this PR.

**Comment by boxdot** [comment]
Looks like I can't push to this PR. I have a fix for register/linking devices.

**Comment by gferon** [comment]
> Looks like I can't push to this PR. I have a fix for register/linking devices.

Thanks! I'll invite you as a collaborator to the repo and then it should work.

File: presage/src/manager/confirmation.rs [hunk 3]
```diff
@@ -104,41 +114,45 @@ impl<S: Store> Manager<S, Confirmation> {
...
-                    capabilities: DeviceCapabilities {
-                        gv2: true,
-                        gv1_migration: true,
-                        ..Default::default()
-                    },
+                    name: None,
+                    capabilities: DeviceCapabilities::default(),
                 },
+                &mut self.store.aci_protocol_store(),
+                &mut self.store.pni_protocol_store(),
```
[the remark this answers is not in the document]
Comment (boxdot) [reply presage/src/manager/confirmation.rs:123]: This now done in c850f2d1a7f22382ae1c2ea104a7759ab4250f49

File: presage/src/manager/registered.rs (hunk 2 above)
Comment (boxdot) [reply presage/src/manager/registered.rs:942]: This is now documented in the issue #252

**Comment by boxdot** [comment]
Manually tested with gurk. So far did not see any problems. Migration did not work, however I am not sure my v4 store state was correct.

**Comment by Schmiddiii** [comment]
> Manually tested with gurk. So far did not see any problems. Migration did not work, however I am not sure my v4 store state was correct.

Testing with Flare works for me. Migration worked as well.

**Comment by boxdot** [comment]
> > Manually tested with gurk. So far did not see any problems. Migration did not work, however I am not sure my v4 store state was correct.
> 
> Testing with Flare works for me. Migration worked as well.

Now, next step is to extend presage's contacts store with a flag whether to send a pni signuture or not.

**Pull request merged**